File: solutions/java/src/main/java/com/epi/JumpGame.java
package com.epi;

import java.util.Random;

public class JumpGame {
  // @include
  public static boolean canReachEnd(int[] maxAdvanceSteps) {
    int furthestReachSoFar = 0;
    for (int i = 0; i <= furthestReachSoFar &&
                    furthestReachSoFar < maxAdvanceSteps.length - 1;
         ++i) {
      furthestReachSoFar = Math.max(furthestReachSoFar, i + maxAdvanceSteps[i]);
    }
    return furthestReachSoFar >= maxAdvanceSteps.length - 1;
  }
  // @exclude

  private static void smallTest() {
    int[] A = {2, 3, 1, 1, 4};
    assert(canReachEnd(A));
    int[] B = {3, 2, 1, 0, 4};
    assert(!canReachEnd(B));
    int[] C = {3, 2, 1, -10, 4};
    assert(!canReachEnd(C));
    int[] D = {2, 3, -1, -1, 4};
    assert(canReachEnd(D));
    int[] E = {2, 2, -1, -1, 100};
    assert(!canReachEnd(E));
  }

  public static void main(String[] args) {
    smallTest();
    Random r = new Random();
    int n;
    if (args.length == 1) {
      n = Integer.parseInt(args[0]);
    } else {
      n = r.nextInt(1000) + 1;
    }
    int[] maxAdvanceSteps = new int[n];
    for (int i = 0; i < n; i++) {
      maxAdvanceSteps[i] = r.nextInt(10) + 1;
    }
    System.out.println(canReachEnd(maxAdvanceSteps));
  }
}

File: static/solutions/jump-game.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <string>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::max;
using std::random_device;
using std::stoul;
using std::uniform_int_distribution;
using std::vector;

// @include
bool CanReachEnd(const vector<int>& max_advance_steps) {
  int furthest_reach_so_far = 0, last_index = max_advance_steps.size() - 1;
  for (int i = 0;
       i <= furthest_reach_so_far && furthest_reach_so_far < last_index;
       ++i) {
    furthest_reach_so_far =
        max(furthest_reach_so_far, max_advance_steps[i] + i);
  }
  return furthest_reach_so_far >= last_index;
}
// @exclude

void SmallTest() {
  vector<int> max_advance_steps = {2, 3, 1, 1, 4};
  assert(CanReachEnd(max_advance_steps));
  max_advance_steps = {3, 2, 1, 0, 4};
  assert(!CanReachEnd(max_advance_steps));
  max_advance_steps = {3, 2, 1, -10, 4};
  assert(!CanReachEnd(max_advance_steps));
  max_advance_steps = {2, 3, -1, -1, 4};
  assert(CanReachEnd(max_advance_steps));
  max_advance_steps = {2, 2, -1, -1, 100};
  assert(!CanReachEnd(max_advance_steps));
}

int main(int argc, char** argv) {
  SmallTest();
  return 0;
  default_random_engine gen((random_device())());
  size_t n;
  if (argc == 2) {
    n = stoul(argv[1]);
  } else {
    uniform_int_distribution<size_t> dis(1, 1000);
    n = dis(gen);
  }
  uniform_int_distribution<int> A_dis(1, 10);
  vector<int> max_advance_steps;
  generate_n(back_inserter(max_advance_steps), n, [&] { return A_dis(gen); });
  cout << std::boolalpha << CanReachEnd(max_advance_steps) << endl;
  return 0;
}
